File: src/moniker_svc/catalog/loader.py
"""Catalog loader - loads catalog definitions from YAML/JSON files."""

from __future__ import annotations

import logging
from pathlib import Path
from typing import Any

import yaml

from .registry import CatalogRegistry
from .types import (
    AccessPolicy, CatalogNode, ColumnSchema, DataQuality, DataSchema,
    Documentation, Freshness, NodeStatus, Ownership, SLA, SourceBinding, SourceType,
)


logger = logging.getLogger(__name__)


class CatalogLoader:
    """
    Loads catalog definitions from YAML or JSON files.

    File format:
    ```yaml
    market-data:
      display_name: Market Data
      description: Real-time and historical market data
      ownership:
        accountable_owner: jane@firm.com
        data_specialist: team@firm.com
        support_channel: "#market-data"

    market-data/prices/equity:
      display_name: Equity Prices
      source_binding:
        type: snowflake
        config:
          account: acme.us-east-1
          database: MARKET_DATA
          query: "SELECT * FROM PRICES WHERE symbol = '{path}'"
    ```
    """

    def load_file(self, path: str | Path) -> CatalogRegistry:
        """Load catalog from a YAML or JSON file."""
        path = Path(path)

        if not path.exists():
            raise FileNotFoundError(f"Catalog file not found: {path}")

        with open(path, "r", encoding="utf-8") as f:
            if path.suffix in (".yaml", ".yml"):
                data = yaml.safe_load(f)
            else:
                import json
                data = json.load(f)

        return self.load_dict(data or {})

    def load_dict(self, data: dict[str, Any]) -> CatalogRegistry:
        """Load catalog from a dictionary."""
        registry = CatalogRegistry()

        for path, node_data in data.items():
            node = self._parse_node(path, node_data)
            registry.register(node)
            logger.debug(f"Loaded catalog node: {path}")

        logger.info(f"Loaded {len(registry.all_paths())} monikers")
        return registry

    def _parse_node(self, path: str, data: dict[str, Any]) -> CatalogNode:
        """Parse a single catalog node from dictionary."""
        # Parse ownership (including formal governance roles)
        ownership = Ownership()
        if "ownership" in data:
            own_data = data["ownership"]
            ownership = Ownership(
                accountable_owner=own_data.get("accountable_owner"),
                data_specialist=own_data.get("data_specialist"),
                support_channel=own_data.get("support_channel"),
                # Formal governance roles (ADOP, ADS, ADAL)
                adop=own_data.get("adop"),
                ads=own_data.get("ads"),
                adal=own_data.get("adal"),
            )

        # Parse source binding
        source_binding = None
        if "source_binding" in data:
            sb_data = data["source_binding"]
            source_type_str = sb_data.get("type", "").lower()

            try:
                source_type = SourceType(source_type_str)
            except ValueError:
                logger.warning(f"Unknown source type '{source_type_str}' for {path}")
                source_type = SourceType.STATIC

            source_binding = SourceBinding(
                source_type=source_type,
                config=sb_data.get("config", {}),
                schema=sb_data.get("schema"),
                read_only=sb_data.get("read_only", True),
            )

        # Parse tags
        tags = frozenset(data.get("tags", []))

        # Parse data quality
        data_quality = None
        if "data_quality" in data:
            dq_data = data["data_quality"]
            data_quality = DataQuality(
                dq_owner=dq_data.get("dq_owner"),
                quality_score=dq_data.get("quality_score"),
                validation_rules=tuple(dq_data.get("validation_rules", [])),
                known_issues=tuple(dq_data.get("known_issues", [])),
                last_validated=dq_data.get("last_validated"),
            )

        # Parse SLA
        sla = None
        if "sla" in data:
            sla_data = data["sla"]
            sla = SLA(
                freshness=sla_data.get("freshness"),
                availability=sla_data.get("availability"),
                support_hours=sla_data.get("support_hours"),
                escalation_contact=sla_data.get("escalation_contact"),
            )

        # Parse freshness
        freshness = None
        if "freshness" in data:
            fresh_data = data["freshness"]
            freshness = Freshness(
                last_loaded=fresh_data.get("last_loaded"),
                refresh_schedule=fresh_data.get("refresh_schedule"),
                source_system=fresh_data.get("source_system"),
                upstream_dependencies=tuple(fresh_data.get("upstream_dependencies", [])),
            )

        # Parse data schema (AI-readable metadata)
        data_schema = None
        if "schema" in data:
            schema_data = data["schema"]
            columns = []
            for col_data in schema_data.get("columns", []):
                columns.append(ColumnSchema(
                    name=col_data.get("name", ""),
                    data_type=col_data.get("type", "string"),
                    description=col_data.get("description", ""),
                    semantic_type=col_data.get("semantic_type"),
                    example=col_data.get("example"),
                    nullable=col_data.get("nullable", True),
                    primary_key=col_data.get("primary_key", False),
                    foreign_key=col_data.get("foreign_key"),
                ))
            data_schema = DataSchema(
                columns=tuple(columns),
                description=schema_data.get("description", ""),
                semantic_tags=tuple(schema_data.get("semantic_tags", [])),
                primary_key=tuple(schema_data.get("primary_key", [])),
                use_cases=tuple(schema_data.get("use_cases", [])),
                examples=tuple(schema_data.get("examples", [])),
                related_monikers=tuple(schema_data.get("related_monikers", [])),
                granularity=schema_data.get("granularity"),
                typical_row_count=schema_data.get("typical_row_count"),
                update_frequency=schema_data.get("update_frequency"),
            )

        # Parse access policy (query guardrails)
        access_policy = None
        if "access_policy" in data:
            ap_data = data["access_policy"]
            access_policy = AccessPolicy(
                required_segments=tuple(ap_data.get("required_segments", [])),
                min_filters=ap_data.get("min_filters", 0),
                blocked_patterns=tuple(ap_data.get("blocked_patterns", [])),
                max_rows_warn=ap_data.get("max_rows_warn"),
                max_rows_block=ap_data.get("max_rows_block"),
                cardinality_multipliers=tuple(ap_data.get("cardinality_multipliers", [])),
                base_row_count=ap_data.get("base_row_count", 100),
                require_confirmation_above=ap_data.get("require_confirmation_above"),
                denial_message=ap_data.get("denial_message"),
                allowed_roles=tuple(ap_data.get("allowed_roles", [])),
            )

        # Parse documentation links (Confluence, runbooks, etc.)
        documentation = None
        if "documentation" in data:
            doc_data = data["documentation"]
            # Parse additional links as tuple of (name, url) pairs
            additional = doc_data.get("additional", {})
            additional_links = tuple((k, v) for k, v in additional.items()) if additional else ()

            documentation = Documentation(
                glossary_url=doc_data.get("glossary"),
                runbook_url=doc_data.get("runbook"),
                onboarding_url=doc_data.get("onboarding"),
                data_dictionary_url=doc_data.get("data_dictionary"),
                api_docs_url=doc_data.get("api_docs"),
                architecture_url=doc_data.get("architecture"),
                changelog_url=doc_data.get("changelog"),
                contact_url=doc_data.get("contact"),
                additional_links=additional_links,
            )

        # Parse lifecycle status from YAML
        status = NodeStatus.ACTIVE
        if "status" in data:
            try:
                status = NodeStatus(data["status"])
            except ValueError:
                logger.warning(f"Unknown status '{data['status']}' for {path}, defaulting to active")

        return CatalogNode(
            path=path,
            display_name=data.get("display_name", ""),
            description=data.get("description", ""),
            domain=data.get("domain"),
            ownership=ownership,
            source_binding=source_binding,
            data_quality=data_quality,
            sla=sla,
            freshness=freshness,
            data_schema=data_schema,
            access_policy=access_policy,
            documentation=documentation,
            classification=data.get("classification", "internal"),
            tags=tags,
            metadata=data.get("metadata", {}),
            status=status,
            deprecation_message=data.get("deprecation_message"),
            successor=data.get("successor"),
            sunset_deadline=data.get("sunset_deadline"),
            migration_guide_url=data.get("migration_guide_url"),
            is_leaf=source_binding is not None,
        )

    def load_directory(self, directory: str | Path) -> CatalogRegistry:
        """
        Load catalog from all YAML/JSON files in a directory.

        Files are loaded in alphabetical order. Later files can override
        earlier definitions.
        """
        directory = Path(directory)
        registry = CatalogRegistry()

        if not directory.is_dir():
            raise NotADirectoryError(f"Not a directory: {directory}")

        files = sorted(directory.glob("*.yaml")) + sorted(directory.glob("*.yml")) + sorted(directory.glob("*.json"))

        for file_path in files:
            logger.info(f"Loading catalog file: {file_path}")
            file_registry = self.load_file(file_path)
            for node in file_registry.all_nodes():
                registry.register(node)

        return registry


def load_catalog(source: str | Path | dict) -> CatalogRegistry:
    """
    Convenience function to load a catalog.

    Args:
        source: File path, directory path, or dictionary

    Returns:
        CatalogRegistry with loaded nodes
    """
    loader = CatalogLoader()

    if isinstance(source, dict):
        return loader.load_dict(source)

    path = Path(source)
    if path.is_dir():
        return loader.load_directory(path)
    else:
        return loader.load_file(path)

File: resolver-go/internal/catalog/loader.go
package catalog

import (
	"fmt"
	"os"

	"gopkg.in/yaml.v3"
)

// CatalogYAML represents the structure of a catalog YAML file
// The catalog YAML is a flat map of path -> node (no "nodes" wrapper)
type CatalogYAML map[string]*CatalogNodeYAML

// CatalogNodeYAML represents a node in the YAML file
type CatalogNodeYAML struct {
	DisplayName       string             `yaml:"display_name"`
	Description       string             `yaml:"description"`
	Domain            *string            `yaml:"domain"`
	Ownership         *OwnershipYAML     `yaml:"ownership"`
	SourceBinding     *SourceBindingYAML `yaml:"source_binding"`
	AccessPolicy      *AccessPolicyYAML  `yaml:"access_policy"`
	Classification    string             `yaml:"classification"`
	DataAssuranceTier *int               `yaml:"data_assurance_tier"`
	Tags              []string           `yaml:"tags"`
	Status            string             `yaml:"status"`
	IsLeaf            bool               `yaml:"is_leaf"`
	Successor         *string            `yaml:"successor"`
}

// OwnershipYAML represents ownership in YAML
type OwnershipYAML struct {
	AccountableOwner *string `yaml:"accountable_owner"`
	DataSpecialist   *string `yaml:"data_specialist"`
	SupportChannel   *string `yaml:"support_channel"`
	ADOP             *string `yaml:"adop"`
	ADS              *string `yaml:"ads"`
	ADAL             *string `yaml:"adal"`
	ADOPName         *string `yaml:"adop_name"`
	ADSName          *string `yaml:"ads_name"`
	ADALName         *string `yaml:"adal_name"`
	UI               *string `yaml:"ui"`
}

// SourceBindingYAML represents a source binding in YAML
type SourceBindingYAML struct {
	Type              string                 `yaml:"type"`
	Config            map[string]interface{} `yaml:"config"`
	AllowedOperations []string               `yaml:"allowed_operations"`
	Schema            map[string]interface{} `yaml:"schema"`
	ReadOnly          *bool                  `yaml:"read_only"`
}

// AccessPolicyYAML represents access policy in YAML
type AccessPolicyYAML struct {
	RequiredSegments       []int    `yaml:"required_segments"`
	MinFilters             *int     `yaml:"min_filters"`
	BlockedPatterns        []string `yaml:"blocked_patterns"`
	MaxRowsWarn            *int     `yaml:"max_rows_warn"`
	MaxRowsBlock           *int     `yaml:"max_rows_block"`
	CardinalityMultipliers []int    `yaml:"cardinality_multipliers"`
	BaseRowCount           *int     `yaml:"base_row_count"`
	DenialMessage          *string  `yaml:"denial_message"`
}

// LoadCatalog loads a catalog from a YAML file
func LoadCatalog(path string) ([]*CatalogNode, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read catalog file: %w", err)
	}

	var catalogYAML CatalogYAML
	if err := yaml.Unmarshal(data, &catalogYAML); err != nil {
		return nil, fmt.Errorf("parse catalog YAML: %w", err)
	}

	nodes := make([]*CatalogNode, 0, len(catalogYAML))
	for path, nodeYAML := range catalogYAML {
		if nodeYAML != nil {
			node := convertYAMLToNode(path, nodeYAML)
			nodes = append(nodes, node)
		}
	}

	return nodes, nil
}

func convertYAMLToNode(path string, yaml *CatalogNodeYAML) *CatalogNode {
	node := &CatalogNode{
		Path:              path,
		DisplayName:       yaml.DisplayName,
		Description:       yaml.Description,
		Domain:            yaml.Domain,
		Classification:    yaml.Classification,
		DataAssuranceTier: yaml.DataAssuranceTier,
		Tags:              yaml.Tags,
		IsLeaf:            yaml.IsLeaf,
		Successor:         yaml.Successor,
	}

	// Set default classification
	if node.Classification == "" {
		node.Classification = "internal"
	}

	// Validate data assurance tier
	if node.DataAssuranceTier != nil {
		tier := *node.DataAssuranceTier
		if tier < 1 || tier > 3 {
			// Log warning but don't fail - be lenient in Go implementation
			fmt.Printf("Warning: data_assurance_tier must be 1, 2, or 3 for node '%s', got: %d\n", path, tier)
			node.DataAssuranceTier = nil // Ignore invalid tier
		}
	}

	// Parse status
	if yaml.Status != "" {
		node.Status = NodeStatus(yaml.Status)
	} else {
		node.Status = NodeStatusActive
	}

	// Convert ownership
	if yaml.Ownership != nil {
		node.Ownership = &Ownership{
			AccountableOwner: yaml.Ownership.AccountableOwner,
			DataSpecialist:   yaml.Ownership.DataSpecialist,
			SupportChannel:   yaml.Ownership.SupportChannel,
			ADOP:             yaml.Ownership.ADOP,
			ADS:              yaml.Ownership.ADS,
			ADAL:             yaml.Ownership.ADAL,
			ADOPName:         yaml.Ownership.ADOPName,
			ADSName:          yaml.Ownership.ADSName,
			ADALName:         yaml.Ownership.ADALName,
			UI:               yaml.Ownership.UI,
		}
	}

	// Convert source binding
	if yaml.SourceBinding != nil {
		readOnly := true
		if yaml.SourceBinding.ReadOnly != nil {
			readOnly = *yaml.SourceBinding.ReadOnly
		}

		node.SourceBinding = &SourceBinding{
			SourceType:        SourceType(yaml.SourceBinding.Type),
			Config:            yaml.SourceBinding.Config,
			AllowedOperations: yaml.SourceBinding.AllowedOperations,
			Schema:            yaml.SourceBinding.Schema,
			ReadOnly:          readOnly,
		}
	}

	// Convert access policy
	if yaml.AccessPolicy != nil {
		baseRowCount := 100
		if yaml.AccessPolicy.BaseRowCount != nil {
			baseRowCount = *yaml.AccessPolicy.BaseRowCount
		}

		node.AccessPolicy = &AccessPolicy{
			RequiredSegments:       yaml.AccessPolicy.RequiredSegments,
			MinFilters:             0,
			BlockedPatterns:        yaml.AccessPolicy.BlockedPatterns,
			MaxRowsWarn:            yaml.AccessPolicy.MaxRowsWarn,
			MaxRowsBlock:           yaml.AccessPolicy.MaxRowsBlock,
			CardinalityMultipliers: yaml.AccessPolicy.CardinalityMultipliers,
			BaseRowCount:           baseRowCount,
			DenialMessage:          yaml.AccessPolicy.DenialMessage,
		}

		if yaml.AccessPolicy.MinFilters != nil {
			node.AccessPolicy.MinFilters = *yaml.AccessPolicy.MinFilters
		}
	}

	return node
}
